AEP/aep-00002.md

```diff
@@ -0,0 +1,158 @@
+---
+AEP: 2
+Title: ArmoniK APIs Definition Guidelines
+Author: Jérôme Gurhem <jgurhem@aneo.fr>
+Status: Draft
+Type: Process
+Content-Type: text/markdown
+Created: 2023-09-14
+---
+
+
+# Abstract
+
+This AEP describes the guidelines that should be followed when updating ArmoniK APIs.
+This mainly pertains to gRPC APIs that are defined in ArmoniK.Api.
+
+# Motivation
+
+In order to create stable API, their development should have guidelines to make sure that they have expected qualities.
+
+# Rationale
+
+Current APIs are not following any guidelines and this led to API which are complicated, not uniform and does not follow any style guide such as [Google Style Guide](https://protobuf.dev/programming-guides/style/).
+This AEP provides guidelines and suggests tools that can be used to ensure that the guidelines are correctly implemented.
+
+# Specification
+
+## APIs should be easy and simple to use
+
+One of the first principles in software development, the KISS principle, should be used to define these APIs. We want API to be as simple possible with the least complexity. This will diminish the difficulty to start using ArmoniK for other developers.
+
+## APIs should have comments
+
+Service, Messages, Enums and RPC methods should be commented so that documentation can be automatically generated.
+These comments should also include noteworthy implementation details such as what is the intended behavior and the expected result.
+
+## Use linter to validate proto files
+
+A protobuf linter must be used to validate that the proto files are properly written. We recommend  [Buf Linter](https://buf.build/docs/lint/overview).
+The following configuration could be used.
+
+```yaml
+version: v1
+lint:
+  use:
+    - DEFAULT
+    - COMMENTS
+```
+
+It includes the following rules:
+
+- Best practices suggest that request and response message for every RPC should be unique as changing one of these message impact every RPC that uses it.
+It is recommended to have a wrapper message for RPC request and response types.
+- Files should declare a package.
+- Files should be in a directory according to their package name and files in the same directory should have the same package.
+- Services and Messages should be named appropriately.
+- Enums should have itself as prefix.
+- Only leading comments are considered by Buf, so proto files should include leading comments for everything.
+
+See [Buf Linter Rules](https://buf.build/docs/lint/rules) for an extensive description of each rule used.
+
+## Avoid using `oneof`
```

Comment:
Maybe add a paragraph about when it is unavoidable to have oneof. I would say when you absolutely need a oneof, you should have a message that consist entirely of a single oneof, nothing else:

```proto
// Avoid this
message MyMessage {
  string id = 0;
  oneof foo {
    string name = 1;
    int id = 2;
  }
  oneof bar {
    string name = 3;
    int id = 4;
  }
}


// Do this instead
message MyMessage {
  message Foo {
    oneof foo {
      string name = 0;
      int id = 1;
    }
  }
  message Bar {
    oneof bar {
      string name = 0;
      int id = 1;
    }
  }

  string id = 0;
  Foo foo = 1;
  Bar bar = 2;
}
```